go/kbfs/simplefs/downloads.go

```diff
@@ -0,0 +1,353 @@
+	}
+	parentDir := m.getDownloadDir()
+	if err = os.MkdirAll(parentDir, 0700); err != nil {
+		return "", err
+	}
+	destPathBase := filepath.Join(parentDir, filename)
+	destPath := destPathBase
+	for suffix := 1; ; suffix++ {
+		_, err := os.Stat(destPath)
+		if os.IsNotExist(err) {
+			break
+		}
+		if err != nil {
+			return "", err
+		}
+		destPath = fmt.Sprintf("%s (%d)", destPathBase, suffix)
+	}
+	// could race but it should be rare enough so fine
+
+	err = os.Rename(srcPath, destPath)
+	if err != nil {
+		// Rename failed; assume it's a cross-partition move.
+		opid, err := m.k.SimpleFSMakeOpid(ctx)
+		if err != nil {
+			return "", err
+		}
+		err = m.k.SimpleFSMove(ctx, keybase1.SimpleFSMoveArg{
+			OpID: opid,
+			Src:  keybase1.NewPathWithLocal(srcPath),
+			Dest: keybase1.NewPathWithLocal(destPath),
+		})
+		if err != nil {
+			return "", err
+		}
+		err = m.k.SimpleFSWait(ctx, opid)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return destPath, nil
+}
+
+func (m *downloadManager) waitForDownload(ctx context.Context,
+	downloadID string, downloadPath string, done func(error)) {
+	d, err := m.getDownload(downloadID)
+	if err != nil {
+		done(err)
+		return
+	}
+	err = m.k.SimpleFSWait(ctx, d.opid)
+	if err != nil {
+		done(err)
+		return
+	}
+
+	var localPath string
+	if d.info.IsRegularDownload {
+		localPath, err = m.moveToDownloadFolder(ctx, downloadPath, d.info.Filename)
+		if err != nil {
+			done(err)
+			return
+		}
+	} else {
+		localPath = downloadPath
+	}
+	if err = m.updateDownload(downloadID, func(d download) download {
+		d.state.LocalPath = localPath
+		return d
+	}); err != nil {
+		done(err)
+		return
+	}
+	done(nil)
+}
+
+func (m *downloadManager) startDownload(
+	ctx context.Context, arg keybase1.SimpleFSStartDownloadArg) (
+	downloadID string, err error) {
+	opid, err := m.k.SimpleFSMakeOpid(ctx)
+	if err != nil {
+		return "", err
+	}
+	downloadID = strconv.FormatInt(time.Now().UnixNano(), 16)
+	downloadPath, filename, err := m.getDownloadPath(ctx, arg.Path, downloadID)
+	if err != nil {
+		return "", err
+	}
+	// TODO for dirs maybe we want zip instead?
+	err = m.k.SimpleFSCopyRecursive(ctx, keybase1.SimpleFSCopyRecursiveArg{
+		OpID: opid,
+		Src:  keybase1.NewPathWithKbfs(arg.Path),
+		Dest: keybase1.NewPathWithLocal(downloadPath),
+	})
+	if err != nil {
+		return "", err
+	}
+
+	func() {
+		defer m.publisher.DownloadStatusChanged()
+		m.lock.Lock()
+		defer m.lock.Unlock()
+		m.downloads[downloadID] = download{
+			info: keybase1.DownloadInfo{
+				DownloadID:        downloadID,
+				Path:              arg.Path,
+				Filename:          filename,
+				StartTime:         keybase1.ToTime(time.Now()),
+				IsRegularDownload: arg.IsRegularDownload,
+			},
+			opid: opid,
+			state: keybase1.DownloadState{
+				DownloadID: downloadID,
+			},
+		}
+	}()
+
+	bgCtx, cancelBtCtx := m.makeContext()
+	done := func(err error) {
+		_ = m.updateDownload(downloadID, func(d download) download {
```

Comment:
I'm not quite sure -- could this function be called twice (once by the monitor and once by the waiter) and race to update `d.state`?  Not sure if that would cause problems or not, but might trigger a data race test failure if this is ever tested.

Reply:
They won't really race they way `-race` detects since `updateDownload` locks. Did you mean we'd get in correct result if they update in unfortunate order? It doesn't seem possible to me as all conditions are checked at the beginning, but please let me know if I missed something!

Reply:
Ah ok, I missed that `updateDownload()` locks.  Thanks.